pytorch_lightning/utilities/fetching.py

```diff
@@ -35,34 +38,60 @@
 class AbstractDataFetcher(ABC):
 
     """
-    This class is used to control batch fetching flow.
+    This based class should be used to implement a fault tolerant ``DataFetcher``.
+    It is required to override the ``fetching_function`` with fetching logic.
+    Example::
+        class SimpleDataFetcher(AbstractDataFetcher):
+            def fetching_function(self):
+                while True:
+                    try:
+                        yield next(self.dataloader_iter), False
+                    except StopIteration:
+                        yield None, True
     """
 
     @abstractmethod
     def fetching_function(self) -> Generator:
-        pass
+        """Override with your own fetching logic."""
 
     def __init__(
         self,
         prefetch_batches: int = 0,
     ) -> None:
-        if not isinstance(prefetch_batches, int) or (isinstance(prefetch_batches, int) and prefetch_batches < 0):
+        if prefetch_batches < 0:
             raise MisconfigurationException("`prefetch_batches` should at least be 0.")
 
         self.prefetch_batches = prefetch_batches + 1
 
         self.dataloader: Optional[Iterable] = None
         self.dataloader_iter: Optional[Iterator] = None
 
+        self.stage: Optional[str]
+        self.batch_to_device: Optional[Callable]
+        self.profiler: "Optional[pl.profiler.base.BaseProfiler]"
+
         self.batches: List
         self.fetched: int
         self.done: bool
 
         self.reset()
 
-    def setup(self, dataloader: DataLoader, **kwargs) -> None:
+    def setup(
+        self,
+        dataloader: Iterable,
+        stage: Optional[str] = None,
+        batch_to_device: Optional[Callable] = None,
+        profiler: "Optional[pl.profiler.base.BaseProfiler]" = None,
+    ) -> None:
         self._add_capture_metadata_collate(dataloader)
+
         self.dataloader = dataloader
+        self.stage = stage
+        self.batch_to_device = batch_to_device
+        self.profiler = profiler
```

Comment:
are these not available at the time of creating the fetcher?

Reply:
what about creating the fetcher in on_trainer_init?

Reply:
Once the DataFetcher is exposed through the Trainer, they won't be available at creation time. 

I prefer the setup function as it is more consistent with other objects we have such as LightningModule, DataModule.